lib/puppet/module/task.rb

```diff
@@ -33,54 +112,75 @@ def self.is_tasks_executable_filename?(name)
         .keep_if { |f| is_tasks_filename?(f) }
-        .group_by { |f| task_name_from_path(f) }
-        .map { |task, files| new_with_files(pup_module, task, files) }
-    end
 
-    attr_reader :name, :module, :metadata_file, :files
+      module_executables = task_files.reject(&method(:is_tasks_metadata_filename?)).map.to_a
 
-    def initialize(pup_module, task_name, files, metadata_file = nil)
-      if !Puppet::Module::Task.is_task_name?(task_name)
-        raise InvalidName, _("Task names must start with a lowercase letter and be composed of only lowercase letters, numbers, and underscores")
+      tasks = task_files.group_by { |f| task_name_from_path(f) }
+
+      tasks.map do |task, executables|
+        new_with_files(pup_module, task, executables, module_executables)
       end
+    end
 
-      all_files = metadata_file.nil? ? files : files + [metadata_file]
-      all_files.each do |f|
-        if !f.start_with?(pup_module.tasks_directory)
-          msg = _("The file '%{path}' is not located in the %{module_name} module's tasks directory") %
-                       {path: f.to_s, module_name: pup_module.name}
+    attr_reader :name, :module, :metadata_file, :metadata, :files
 
-          # we can include some extra context for the log message:
-          Puppet.err(msg + " (#{pup_module.tasks_directory})")
-          raise InvalidFile, msg
-        end
+    # file paths must be relative to the modules task directory
+    def initialize(pup_module, task_name, files, metadata_file = nil, module_executables = nil)
+      if !Puppet::Module::Task.is_task_name?(task_name)
+        raise InvalidName, _("Task names must start with a lowercase letter and be composed of only lowercase letters, numbers, and underscores")
       end
 
       name = task_name == "init" ? pup_module.name : "#{pup_module.name}::#{task_name}"
 
       @module = pup_module
       @name = name
-      @metadata_file = metadata_file if metadata_file
+      @metadata_file = metadata_file
       @files = files
+      @module_executables = module_executables || []
+    end
+
+    def read_metadata(file)
+      File.open(file) { |fh| JSON.load(fh) } if file
+    rescue SystemCallError, IOError => err
+      msg = _("Error reading metadata: %{message}" % {message: err.message})
+      raise InvalidMetadata.new(msg, 'puppet.tasks/unreadable-metadata')
+    rescue JSON::ParserError => err
+      raise InvalidMetadata.new(err.message, 'puppet.tasks/unparseable-metadata')
+    end
+
+    def metadata
+      @metadata ||= read_metadata(@metadata_file)
+    end
+
+    def implementations
+      return @implementations if @implementations
+      executables = (@files + @module_executables).uniq
+      @implementations = self.class.find_implementations(@name, @module.tasks_directory, metadata, executables)
+    end
+
+    def files
+      implementations.map {|imp| imp.select {|k,v| ['name', 'path'].include?(k) } }
```

Comment:
Each entry should *always* have both a `name` and `path`, correct? In that case, this might be clearer as just `{'name' => imp['name'], 'path' => imp['path']}`.